share/search/indexing.py

```diff
@@ -0,0 +1,298 @@
+            self.message.payload,
+        ))
+
+    def __iter__(self):
+        raise NotImplementedError
+
+    def __next__(self):
+        raise NotImplementedError
+
+    def _to_model(self, name):
+        name = name.lower()
+
+        if name.startswith('share.'):
+            model = apps.get_model(name)
+        else:
+            model = apps.get_model('share', name)
+
+        if not issubclass(model, ShareObject):
+            raise ValueError('Invalid model "{!r}"'.format(model))
+
+        # Kinda a hack, grab the first non-abstract version of a typed model
+        if model._meta.concrete_model.__subclasses__():
+            return model._meta.concrete_model.__subclasses__()[0]
+
+        return model
+
+
+class V0Message(IndexableMessage):
+    """
+    {
+        "<model_name>": [id1, id2, id3...]
+    }
+    """
+    PROTOCOL_VERSION = 0
+
+    @property
+    def model(self):
+        return self._to_model(next(iter(self.message.payload.keys())))
+
+    def __init__(self, message):
+        super().__init__(message)
+
+        self.message.payload.pop('version', None)
+
+        if len(self.message.payload.keys()) > 1:
+            raise self.malphormed('Multiple models')
+
+    def __len__(self):
+        return len(next(iter(self.message.payload.values())))
+
+    def __iter__(self):
+        return iter(next(iter(self.message.payload.values())))
+
+
+class V1Message(IndexableMessage):
+    """
+    {
+        "version": 1,
+        "model": "<model_name>",
+        "ids": [id1, id2, id3...]
+    }
+    """
+    PROTOCOL_VERSION = 1
+
+    @property
+    def model(self):
+        return self._to_model(self.message.payload['model'])
+
+    def __iter__(self):
+        return iter(self.message.payload['ids'])
+
+    def __len__(self):
+        return len(self.message.payload['ids'])
+
+
+class MessageFlattener:
+
+    def __init__(self, messages):
+        self.acked = []
+        self.pending = []
+        self.requeued = []
+
+        self.current = None
+        self.messages = collections.deque(messages)
+        self.buffer = collections.deque()
+
+    def ack_pending(self):
+        if not self.pending:
+            return
+
+        for message in self.pending:
+            message.message.ack()
+            self.acked.append(message)
+
+        self.pending.clear()
+
+    def requeue_pending(self):
+        if not self.pending:
+            return
+
+        for message in self.pending:
+            message.message.requeue()
+            self.requeued.append(message)
+
+        self.pending.clear()
+
+    def reset_pending(self):
+        if not self.pending:
+            return
+
+        for message in self.pending:
+            self.messages.append(message)
+
+        self.current = None
+        self.pending.clear()
+
+    def __len__(self):
+        return sum([len(msg) for msg in self.messages], 0)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.current is None:
```

Comment:
It forces a "look ahead" so that we're always one ahead of the internal buffer. So as soon as the final element of a list is consumed, it will be pushed into pending rather than on the next call. 

I moved the initial call into `__init__` because that make more sense 👍